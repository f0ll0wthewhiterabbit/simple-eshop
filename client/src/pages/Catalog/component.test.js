import React from 'react'
import { shallow } from 'enzyme'
import { List, Record } from 'immutable'

import CatalogPage from './component'
import findByTestAttr from '../../utils/findByTestAttr'

describe('CatalogPage component', () => {
  let wrapper
  const initialProps = {
    products: List(),
    error: null,
    isLoading: false,
    fetchProductsRequest: jest.fn(),
  }
  const testProduct1 = Record({ _id: '1', title: 'First', imageName: 'imageName1', price: 77 })()
  const testProduct2 = Record({ _id: '2', title: 'Second', imageName: 'imageName2', price: 88 })()
  const testProduct3 = Record({ _id: '3', title: 'Third', imageName: 'imageName3', price: 99 })()
  const testProductsList = List([testProduct1, testProduct2, testProduct3])
  const dataTestErrorMessage = 'errorMessage'
  const dataTestLoader = 'loader'
  const dataTestProductCard = 'productCard'

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<CatalogPage {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render correct text in error message', () => {
    wrapper = generateWrapper({
      products: testProductsList,
      error: 'test error message',
    })
    const errorMessage = findByTestAttr(wrapper, dataTestErrorMessage)

    expect(errorMessage.text()).toBe('Sorry, something went wrong...')
  })

  it('should render error message after error', () => {
    wrapper = generateWrapper({
      products: testProductsList,
      error: 'test error message',
    })
    const errorMessage = findByTestAttr(wrapper, dataTestErrorMessage)

    expect(errorMessage).toHaveLength(1)
  })

  it('should render loader', () => {
    wrapper = generateWrapper({
      products: testProductsList,
      isLoading: true,
    })
    const loader = findByTestAttr(wrapper, dataTestLoader)

    expect(loader).toHaveLength(1)
  })

  it('should render correct amount of ProductCards', () => {
    wrapper = generateWrapper({
      products: testProductsList,
    })
    const productCard = findByTestAttr(wrapper, dataTestProductCard)

    expect(productCard).toHaveLength(testProductsList.size)
  })
})
