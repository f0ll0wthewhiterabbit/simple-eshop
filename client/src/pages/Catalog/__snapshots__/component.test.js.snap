// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogPage component should render correctly 1`] = `
<Styled(WithStyles(ForwardRef(Container)))
  maxWidth="lg"
>
  <Styled(WithStyles(ForwardRef(Typography)))
    align="center"
    component="h1"
    variant="h4"
  >
    Catalog
  </Styled(WithStyles(ForwardRef(Typography)))>
  <Connect(FilterSelect) />
  <styled.div>
    <WithStyles(ForwardRef(Typography))
      align="center"
      data-test="errorMessage"
      variant="body1"
    >
      Sorry, there are no products yet
    </WithStyles(ForwardRef(Typography))>
  </styled.div>
</Styled(WithStyles(ForwardRef(Container)))>
`;
